src/stages/stage5/spells/artificial_lightning.c

```diff
@@ -10,148 +10,155 @@
-			.pos = p->pos,
-			.color = RGBA(0.1, 0.1, 0.6, 0.0),
-			.timeout = 15,
-			.draw_rule = Fade,
+			.pos = boss->pos,
+			.color = RGBA(0.05, 0.05, 0.3, 0),
+			.draw_rule = pdraw_timeout_fade(0.5, 0),
+			.timeout = l/s,
+			.move = move_linear(-s * n),
 		);
 	}
+}
+
+TASK(zigzag_move, { BoxedProjectile p; cmplx move_arg; }) {
+	Projectile *p = TASK_BIND(ARGS.p);
+
+	for(int time = 0;; time++, YIELD) {
+		// TODO: make this better
+		p->pos = p->pos0 + (abs(((2 * time) % 50) - 50 / 2) * I + time) * 2 * ARGS.move_arg;
 
-	return ACTION_NONE;
+		if(time % 2 == 0) {
+			PARTICLE(
+				.sprite = "lightningball",
+				.pos = p->pos,
+				.color = RGBA(0.1, 0.1, 0.6, 0.0),
+				.timeout = 15,
+				.draw_rule = pdraw_timeout_fade(0.5, 0),
+			);
+		}
+	}
 }
 
-static int lightning_slave(Enemy *e, int t) {
-	if(t < 0)
-		return 1;
-	if(t > 200)
-		return ACTION_DESTROY;
+TASK(zigzag_shoot, { BoxedBoss boss; } ) {
+	Boss *boss = TASK_BIND(ARGS.boss);
+	for(int time = 0;; time++, YIELD) {
+		if(time > 0 && !(time % 100)) {
+			int count = difficulty_value(7, 7, 7, 9);
+			for(int i = 0; i < count; i++) {
+				Projectile *p = PROJECTILE(
+					.proto = pp_bigball,
+					.pos = boss->pos,
+					.color = RGBA(0.5, 0.1, 1.0, 0.0),
+				);
+				INVOKE_SUBTASK(zigzag_move, ENT_BOX(p), cdir(M_TAU / count * i) * cnormalize(global.plr.pos - boss->pos));
+			}
+			play_sfx("redirect");
+			play_sfx("shot_special1");
+		}
+	}
+}
+
+TASK(lightning_slave_move, { BoxedEnemy e; cmplx move_arg; }) {
+	Enemy *e = TASK_BIND(ARGS.e);
+	cmplx move = ARGS.move_arg;
 
-	TIMER(&t);
+	for(int time = 0;; time++, YIELD) {
+		e->move = move_linear(move);
+		if(time%20 == 0) {
+			move *= cdir(0.25 + 0.25 * rng_real() * M_PI);
+		}
+	}
+}
 
-	e->pos += e->args[0];
+TASK(lightning_slave, { cmplx pos; cmplx move_arg; }) {
+	// TODO: get rid of ENEMY_IMMUNE
+	Enemy *e = TASK_BIND(create_enemy1c(ARGS.pos, ENEMY_IMMUNE, NULL, NULL, 0));
 
-	FROM_TO(0,200,20)
-		e->args[0] *= cexp(I * (0.25 + 0.25 * frand() * M_PI));
+	INVOKE_TASK(lightning_slave_move, ENT_BOX(e), ARGS.move_arg);
 
-	FROM_TO(0, 200, 3)
-		if(cabs(e->pos-global.plr.pos) > 60) {
-			Color *clr = RGBA(1-1/(1+0.01*_i), 0.5-0.01*_i, 1, 0);
+	for(int x = 0; x < 67; x += WAIT(3)) {
+		if(cabs(e->pos - global.plr.pos) > 60) {
+			Color *clr = RGBA(1 - 1 / (1 + 0.01 * x), 0.5 - 0.01 * x, 1, 0);
 
 			Projectile *p = PROJECTILE(
 				.proto = pp_wave,
 				.pos = e->pos,
 				.color = clr,
-				.rule = asymptotic,
-				.args = {
-					0.75*e->args[0]/cabs(e->args[0])*I,
-					10
-				},
+				// TODO: this doesn't quite work right
+				.move = move_asymptotic_simple(0.75 * ARGS.move_arg / cabs(ARGS.move_arg) * I, 10),
 			);
 
 			if(projectile_in_viewport(p)) {
 				for(int i = 0; i < 3; ++i) {
-					tsrand_fill(2);
-					iku_lightning_particle(p->pos + 5 * afrand(0) * cexp(I*M_PI*2*afrand(1)), 0);
+					RNG_ARRAY(rand, 2);
+					iku_lightning_particle(p->pos + 5 * vrng_sreal(rand[0]) * cdir(M_TAU * vrng_sreal(rand[1])));
```

Comment:
`vrng_sreal` -> `vrng_real` although here it doesn’t affect the result.

Reply:
Actually just replace that whole `cdir(M_TAU * vrng_sreal(rand[1])))` with `vrng_dir(rand[1])`.